x/ibc/06-solomachine/types/client_state.go

```diff
@@ -0,0 +1,438 @@
+	return Initialize(msg.GetClientID(), msg.ConsensusState)
+}
+
+// Initialize creates an unfrozen client with the initial consensus state
+func Initialize(id string, consensusState ConsensusState) (ClientState, error) {
+	return ClientState{
+		ID:             id,
+		Frozen:         false,
+		ConsensusState: consensusState,
+	}, nil
+}
+
+// NewClientState creates a new ClientState instance.
+func NewClientState(id string, consensusState ConsensusState) ClientState {
+	return ClientState{
+		ID:             id,
+		Frozen:         false,
+		ConsensusState: consensusState,
+	}
+}
+
+// GetID returns the solo machine client state identifier.
+func (cs ClientState) GetID() string {
+	return cs.ID
+}
+
+// GetChainID returns an empty string.
+func (cs ClientState) GetChainID() string {
+	return ""
+}
+
+// ClientType is Solo Machine.
+func (cs ClientState) ClientType() clientexported.ClientType {
+	return clientexported.SoloMachine
+}
+
+// GetLatestHeight returns the latest sequence number.
+func (cs ClientState) GetLatestHeight() uint64 {
+	return cs.ConsensusState.Sequence
+}
+
+// IsFrozen returns true if the client is frozen
+func (cs ClientState) IsFrozen() bool {
+	return cs.Frozen
+}
+
+// Validate performs basic validation of the client state fields.
+func (cs ClientState) Validate() error {
+	if err := host.ClientIdentifierValidator(cs.ID); err != nil {
+		return err
+	}
+	return cs.ConsensusState.ValidateBasic()
+}
+
+// VerifyClientConsensusState verifies a proof of the consensus state of the
+// Solo Machine client stored on the target machine.
+func (cs ClientState) VerifyClientConsensusState(
+	store sdk.KVStore,
+	cdc *codec.Codec,
+	root commitmentexported.Root,
+	_ uint64,
+	counterpartyClientIdentifier string,
+	consensusHeight uint64,
+	prefix commitmentexported.Prefix,
+	proof commitmentexported.Proof,
+	consensusState clientexported.ConsensusState,
+) error {
+	clientPrefixedPath := "clients/" + counterpartyClientIdentifier + "/" + host.ConsensusStatePath(consensusHeight)
+	path, err := commitmenttypes.ApplyPrefix(prefix, clientPrefixedPath)
+	if err != nil {
+		return err
+	}
+
+	if cs.IsFrozen() {
+		return clienttypes.ErrClientFrozen
+	}
+
+	// cast the proof to a signature proof
+	signatureProof, ok := proof.(commitmenttypes.SignatureProof)
+	if !ok {
+		return sdkerrors.Wrapf(clienttypes.ErrInvalidClientType, "proof type %T is not type SignatureProof", proof)
+	}
+
+	bz, err := cdc.MarshalBinaryBare(consensusState)
+	if err != nil {
+		return err
+	}
+
+	// value = sequence + path + consensus state
+	value := append(
+		combineSequenceAndPath(cs.ConsensusState.Sequence, path),
+		bz...,
+	)
+	if !cs.ConsensusState.PubKey.VerifyBytes(value, signatureProof.Signature) {
+		return sdkerrors.Wrap(clienttypes.ErrFailedClientConsensusStateVerification, "failed to verify proof against current public key, sequence, and consensus state")
+	}
+
+	cs.ConsensusState.Sequence++
+	setClientState(store, cs)
```

Comment:
These set functions **cannot** just set the latest consensus state in the client state. They should also be storing the consensus state where the 02-client keeper would.

store.Set(`host.KeyConsensusState(sequence), consensusState)`

This is necessary so past misbehavior can be processed correctly

Reply:
should `sequence` be `consensusState.Sequence`, `height` (as passed into verify func) or `clientState.LatestSequence`?

Reply:
seems like it should be the `height` or in this case `sequence` as passed in by the verify func? This should be equal to `consensusState.Sequence` since the consensusState is assumed to be store by its sequence. Should I remove `Sequence` from consensusState and just leave consensusState as the pubkey?

Reply:
Hmm, I think the consensusState should be stored at the clientState.LatestSequence + 1 like so:

```go
clientState.LatestSequence += 1
store.Set(host.KeyConsensusState(clientState.LatestSequence), consensusState)
store.Set(host.KeyClientState(), clientState)
```

this is because the consensusState is retrieved from the clientKeeper using the provided sequence/height. So the sequence is used for this proof which doesn't result in a pubkey change so the sequence + 1 should contain the same consensusState. Not incrementing it in this fashion would result in an empty consensusState

Reply:
Consensus state for solo machines is not stored per-sequence. I realise this nomenclature is confusing, maybe it should be changed in the spec - essentially, solo machines "just" have a client state, not a consensus state for each height.

Reply:
I found it confusing that the consensus state is stored as a ClientState field.